language-plutus-core/src/Language/PlutusCore/CBOR.hs

```diff
@@ -272,3 +273,90 @@ instance Serialise UnknownDynamicBuiltinNameError
 instance (Closed uni, uni `Everywhere` Serialise, Serialise ann) => Serialise (InternalTypeError uni ann)
 instance (Closed uni, uni `Everywhere` Serialise, Serialise ann) => Serialise (TypeError uni ann)
 instance (Closed uni, uni `Everywhere` Serialise, Serialise ann) => Serialise (Error uni ann)
+
+
+{- Serialising the unit annotation takes up space: () is converted to
+   the CBOR `null` value, which is encoded as the byte 0xF6.  In
+   typical examples these account for 30% or more of the bytes in a
+   serialised PLC program.  We don't actually need to serialise unit
+   annotations since we know where they're going to appear when we're
+   deserialising, and we know what the value has to be.  The
+   `InvisibleUnit` type below has a `Serialise` instance which
+   takes care of this for us: `serialise (InvisibleUnit () <$ prog)`
+    will serialise a program omitting the unit annotations, and
+   `deserialise` will give us back an `InvisibleUnit`-annotated program
+   which we can convert back to a ()-annotated program with `() <$ ...`.
+   The functions `serialiseProg`, `deserialiseProg` and `deserialiseProgOrFail`
+   do the conversions for us, so that the `InvisibleUnit` type doesn't have
+   to be exposed to uesrs.  These functions should always be used to serialise
+   and deserialise ()-annotated programs.
+
+   A possible problem is that `() <$ prog` has to visit every node in
```

Comment:
You could also run `withMemory` over it and add the memory consumption annotation here. It's gotta happen somewhere.

Reply:
Well, that also takes space. The point here is to make it small for transmission.

Reply:
On the deserilization side only.